worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,334 @@
+	// subscriptions. This allows the queue to be lock less.
+	subscriptionCh   chan subscriptionOpts
+	unsubscriptionCh chan uint64
+}
+
+// New creates a new EventQueue that will use the Stream for events.
+func New(stream Stream, logger Logger) (*EventQueue, error) {
+	queue := &EventQueue{
+		stream:             stream,
+		logger:             logger,
+		subscriptions:      make(map[uint64]*subscription),
+		subscriptionsByNS:  make(map[string][]*eventFilter),
+		subscriptionsAll:   make(map[uint64]struct{}),
+		subscriptionsCount: 0,
+
+		subscriptionCh:   make(chan subscriptionOpts),
+		unsubscriptionCh: make(chan uint64),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &queue.catacomb,
+		Work: queue.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return queue, nil
+}
+
+// Subscribe creates a new subscription to the event queue. Options can be
+// provided to allow filter during the dispatching phase.
+func (q *EventQueue) Subscribe(opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	// Get a new subscription count without using any mutexes.
+	subID := atomic.AddUint64(&q.subscriptionsCount, 1)
+
+	sub := newSubscription(subID, func() { q.unsubscribe(subID) })
+	if err := q.catacomb.Add(sub); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	select {
+	case <-q.catacomb.Dying():
+		return nil, q.catacomb.ErrDying()
+	case q.subscriptionCh <- subscriptionOpts{
+		subscription: sub,
+		opts:         opts,
+	}:
+	}
+
+	return sub, nil
+}
+
+// Kill stops the event queue.
+func (q *EventQueue) Kill() {
+	q.catacomb.Kill(nil)
+}
+
+// Wait waits for the event queue to stop.
+func (q *EventQueue) Wait() error {
+	return q.catacomb.Wait()
+}
+
+func (q *EventQueue) unsubscribe(subscriptionID uint64) {
+	select {
+	case <-q.catacomb.Dying():
+		return
+	case q.unsubscriptionCh <- subscriptionID:
+	}
+}
+
+func (q *EventQueue) loop() error {
+	defer func() {
+		// There isn't a need to close subscriptions directly, as they're tied
+		// to the catacomb via their own tomb.
+
+		q.subscriptions = nil
+		q.subscriptionsByNS = nil
+
+		close(q.subscriptionCh)
+		close(q.unsubscriptionCh)
+	}()
+
+	for {
+		select {
+		case <-q.catacomb.Dying():
+			return q.catacomb.ErrDying()
+
+		case event, ok := <-q.stream.Changes():
+			// If the stream is closed, we expect that a new worker will come
+			// again using the change stream worker infrastructure. In this case
+			// just ignore and close out.
+			if !ok {
+				q.logger.Infof("change stream change channel is closed")
+				return nil
+			}
+
+			subs := q.gatherSubscriptions(event)
+			for _, sub := range subs {
+				// Ensure we check tomb dying as the handling logic is
+				// synchronous and blocking. This is to effectively handle
+				// back pressure on the stream.
+				select {
+				case <-q.catacomb.Dying():
+					return q.catacomb.ErrDying()
+				case sub.signal() <- event:
```

Comment:
only problem with this is that the unsubscribe can't immediately skip this blocking send.

Reply:
OTTOMH, we could use a "subscription active" channel or something that gets closed by unsubscribe and is used similarly to tomb dying in the signal() func.

Reply:
I think this is a reasonable approach.